### DISCUSSION TITLE
updi support with Linux gpio

### DISCUSSION THREAD

**Discussion by gcormier** [avrdudes/avrdude#1456] [Ideas]
Would it be possible to use linuxgpio to bitbang the serial for `PM_UPDI`, as is the case with `PM_ISP`? This might require additional code. Not sure which mapping would relate to the actual UPDI data pin.

ISP Example
```
programmer
  id          = "avrhat_5";
  desc        = "Use the Linux sysfs interface to bitbang GPIO lines";
  type        = "linuxgpio";
  prog_modes  = PM_ISP;
  reset       = 19;
  sck         = 20;
  mosi        = 21;
  miso        = 26;
;

**Comment by mcuee** [comment]
I think this should be a possible enhancement but somebody needs to contribute to write the codes. As for the mapping it is probably quite flexible as you are using GPIO and not SPI/UART/etc.

In the mean time, you may want to use serialupdi or jtag2updi.
1) serialupdi -- low cost and yet with good performance
https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md

2) jtag2updi -- if you have Uno/Nano
https://github.com/ElTangas/jtag2updi

**Comment by MCUdude** [comment]
You should be able to build your own SerialUPDI programmer using the RX and TX lines on your Raspberry Pi Header.
I don't see the reason why anyone would want to use a bit-banged, GPIO-based SerialUPDI programmer when they can use the UARTS pins directly and achieve really high speeds.

**Comment by gcormier** [comment]
> I don't see the reason why

I can solve that one.

https://github.com/gcormier/avr5hat

This allows me to program 5 ISP targets at once. With the requested functionality, the same could be done for UPDI targets.

**Comment by mcuee** [comment]
Just FYI, there are boards with 4 USB to TTL converters with 5V/3/3V/2.5V/1.8V selections (for Vcc, TX and RX pins).

Example: about US$12 from AliExpress, using FT4232HL.
![FT4232_quad_USB_to_TTL](https://user-images.githubusercontent.com/1294740/233993524-ce06fc88-4051-41ac-a75d-9a2ac62c4390.png)

I have another version based on the same FT4232HL with only 5V and 3.3V, but with all the pins available so that it can work with USB to two JTAG (I used it for testing of PR #1324), USB to 4 SPI as well as USB to 4 TTL. It is probably a bit more expensive than the above.
![FT4232HL_quad](https://user-images.githubusercontent.com/1294740/233995747-b2683629-3005-497d-8298-a175534c53db.jpg)

There are also cheaper ones with a USB hub and 4 CH340 USB to TTL converters.

**Comment by MCUdude** [comment]
@dbuchwald was the one that wrote the SerialUPDI implementation, so maybe he can answer if this is a doable thing or not.

**Comment by dbuchwald** [comment]
@MCUdude, this is generally doable, as everything else, but the effort in AVRDUDE is difficult to justify. 

What I would recommend to @gcormier is to look if there are any software based bit-bang UART devices available (see https://www.linuxquestions.org/questions/linux-embedded-and-single-board-computer-78/software-uart-driver-4175524941/ for instance) and use something like that for the mapping of virtual UART devices to GPIO pins. Then he could have AVRDUDE operate as normal by using these UART devices.

Does that make sense, @gcormier?

**Comment by gcormier** [comment]
@dbuchwald Good call. There is indeed one, at https://github.com/adrianomarto/soft_uart

That said, I'm realizing now it might not be as simple as that. As mentioned on [serialupdi guide](https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md), hardware modifications are needed. Or, linuxgpio would need to switch between read/write at sufficient speed which I am rather skeptical of, and would agree that would not be fun to support with avrdude.

That said, I suspect we can close this issue. It is a bit of an edge case, but someone was asking if this board could be used for both ISP and UPDI. I think we found the answer :)

**Comment by mcuee** [comment]
Okay, I will close this as not planned.

In case someone is interested enough and creates a PR, then we can reopen this issue.
